pkg/updater/download_test.go

```diff
@@ -126,39 +255,75 @@ func addToArchive(dir string, tw *tar.Writer, filename string) error {
 	return nil
 }
 
-func TestDownload(t *testing.T) {
-	dir := t.TempDir()
-	server := createTestServer(t, dir)
-	defer server.Close()
-	downloader := newDownloader(server.Client())
-	downloadPath := path.Join(dir, testDownloadDir)
+func (suite *DownloadTestSuite) TestDownload() {
+	t := suite.T()
+	downloader := newDownloader(suite.server.Client())
+	downloadPath := path.Join(suite.dir, testDownloadDir)
 	err := os.MkdirAll(downloadPath, 0755)
 	assert.NoError(t, err)
 
-	pkg := Package{URL: fmt.Sprintf("%s/%s", server.URL, testAgentArchiveFileName), SHA256: agentArchiveHash(t, dir)}
+	pkg := Package{
+		Name:    "datadog-agent",
+		Version: "7.51.0",
+		URL:     fmt.Sprintf("%s/%s", suite.server.URL, testAgentArchiveFileName),
+		SHA256:  agentArchiveHash(t, suite.dir),
+		Size:    agentArchiveSize(t, suite.dir),
+	}
+
+	// Register how much memory we used setting up the test
+	var m runtime.MemStats
+	runtime.ReadMemStats(&m)
+	totalAllocsStart := m.TotalAlloc
+
 	err = downloader.Download(context.Background(), pkg, downloadPath)
+	runtime.ReadMemStats(&m)
+
+	// Read contents of downloadPath
 	assert.NoError(t, err)
 	assert.FileExists(t, path.Join(downloadPath, testAgentFileName))
 	assert.FileExists(t, path.Join(downloadPath, testNestedAgentFileName))
 	assert.FileExists(t, path.Join(downloadPath, testLargeFileName))
 
 	// ensures the full archive or full individual files are not loaded in memory
-	var m runtime.MemStats
-	runtime.ReadMemStats(&m)
-	assert.Less(t, m.TotalAlloc, uint64(testLargeFileSize))
+	xzDictCap := uint64(8428576) // 8 MiB, the default xz dictionary size
+	allocatedMemoryArchive := m.TotalAlloc - totalAllocsStart
+	expectedMemory := 2 * (uint64(testLargeFileSize) + xzDictCap)
```

Comment:
Let's make sure to revert that part, it should stay under `testLargeFileSize`. Keep that file to 20MB or bump it slightly and it shouldn't fail (or we haven't found another issue).

Reply:
[This](https://github.com/DataDog/datadog-agent/pull/22812/files#diff-93a3bc13b016de910d4e39c9c91cba5b8906520cb809bb0e473bc8badcfc0e53R246) `io.Copy` uses >50% of any file it's given. It is called twice as we have two tar extractions, so we use more than `testLargeFileSize` in total.

Reply:
Going back to gz fixed the issue